github/util.go

```diff
@@ -0,0 +1,206 @@
+func validateIdentityFields(ref gitprovider.IdentityRef, expectedDomain string) error {
+	// Make sure the expected domain is used
+	if ref.GetDomain() != expectedDomain {
+		return gitprovider.ErrDomainUnsupported
+	}
+	// Make sure the right type of identityref is used
+	switch ref.GetType() {
+	case gitprovider.IdentityTypeOrganization, gitprovider.IdentityTypeUser:
+		return nil
+	case gitprovider.IdentityTypeSuborganization:
+		return gitprovider.ErrProviderNoSupport
+	}
+	return gitprovider.ErrInvalidArgument
+}
+
+// resolveOrg returns the organization name if ref is an organization, or
+// an empty string if it is an user account.
+func resolveOrg(ref gitprovider.IdentityRef) string {
+	// If the ref is an organization, return its name
+	if ref.GetType() == gitprovider.IdentityTypeOrganization {
+		return ref.GetIdentity()
+	}
+	// If the ref is an user account, return an empty string
+	return ""
+}
+
+func repositoryFromAPI(apiObj *github.Repository, ref gitprovider.RepositoryRef) *gitprovider.Repository {
+	repo := &gitprovider.Repository{
+		RepositoryInfo: gitprovider.FromRepositoryRef(ref),
+		InternalHolder: gitprovider.WithInternal(apiObj),
+	}
+	if apiObj.Description != nil && len(*apiObj.Description) != 0 {
+		repo.Description = apiObj.Description
+	}
+	if apiObj.DefaultBranch != nil && len(*apiObj.DefaultBranch) != 0 {
+		repo.DefaultBranch = apiObj.DefaultBranch
+	}
+	if apiObj.Visibility != nil && len(*apiObj.Visibility) != 0 {
+		// TODO: What should we do if *apiObj.Visibility wouldn't be any of the already-known values?
+		repo.Visibility = gitprovider.RepositoryVisibilityVar(gitprovider.RepositoryVisibility(*apiObj.Visibility))
+	}
+	return repo
+}
+
+func repositoryToAPI(repo *gitprovider.Repository) *github.Repository {
+	apiObj := &github.Repository{
+		Name:          &repo.RepositoryName,
+		Description:   repo.Description,
+		DefaultBranch: repo.DefaultBranch,
+	}
+	if repo.Visibility != nil {
+		apiObj.Visibility = gitprovider.StringVar(string(*repo.Visibility))
+	}
+	return apiObj
+}
+
+func applyRepoCreateOptions(apiObj *github.Repository, opts gitprovider.RepositoryCreateOptions) {
+	apiObj.AutoInit = opts.AutoInit
+	if opts.LicenseTemplate != nil {
+		apiObj.LicenseTemplate = gitprovider.StringVar(string(*opts.LicenseTemplate))
+	}
+}
+
+func organizationFromAPI(apiObj *github.Organization, domain string) *gitprovider.Organization {
+	return &gitprovider.Organization{
+		OrganizationInfo: gitprovider.OrganizationInfo{
+			Domain:       domain,
+			Organization: *apiObj.Login,
+		},
+		InternalHolder: gitprovider.WithInternal(apiObj),
+		Name:           apiObj.Name,
+		Description:    apiObj.Description,
+	}
+}
+
+// handleHTTPError checks the type of err, and returns typed variants of it
+// However, it _always_ keeps the original error too, and just wraps it in a MultiError
+// The consumer must use errors.Is and errors.As to check for equality and get data out of it
+func handleHTTPError(err error) error {
```

Comment:
This can pass through `nil` errors, so I'm for simplifying cases like this:
```
if err != nil {
    return handleHTTPError(err)
}
return nil
```
to just `return handleHTTPError(err)`. Also add a fast-path `if err != nil => return` here to prevent excess allocations.

Reply:
> This can pass through nil errors, so I'm for simplifying cases like this:

yes, done.

> Also add a fast-path if err != nil => return here to prevent excess allocations.

I'm not sure I understand. We want to wrap the error in various ways if set.

Reply:
Sorry, I meant `if err == nil { return nil }` :sweat_smile: so if the error is nil (unset) we don't need to allocate any structs or do type assertion.